### PR TITLE
[IOPID-536] : Migration from cookies to session storage

### PR DESCRIPTION
**Pull request by claudiodall**
## Short description

As discussed in the meeting, we are not allowed to store tokens and user info inside the cookie. The purpose of these fixes is to change the logic where the cookies were intercepted from the header of the request in the middleware. Since the middleware has no chance to get data from session storage, we migrated the logic to the session provider to retrieve the session storage from there.

## List of changes proposed in this pull request

- Removed logic from middleware.
- Added session provider.
- Added IdP (Identity Provider) list selection to /access (SPID) to enable login functionality.
- Updated routes.

## How to test

Try to log in through **/access** or **/logoutInit** using the SPID and check if the token and user object are stored in session storage instead of cookies.

To log in, use hub-spid-login as SP or force it with the following link+token:

http://localhost:3000/validateSession#token=eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6Imp3dF84Njo3NDoxZTozNTphZTphNjpkODo0YjpkYzplOTpmYzo4ZTphMDozNTo2ODpiNSJ9.eyJlbWFpbCI6InBpcHBvQHRlc3QuZW1haWwuaXQiLCJmYW1pbHlfbmFtZSI6InF3ZXJ0eSIsImZpc2NhbF9udW1iZXIiOiJRV1JQUFA4MEEwMUg1MDFGIiwibmFtZSI6InBpcHBvIiwiZnJvbV9hYSI6ZmFsc2UsInVpZCI6IjgzODQzODY0LWYzYzAtNGRlZi1iYWRiLTdmMTk3NDcxYjcyZSIsImxldmVsIjoiTDIiLCJpYXQiOjE2ODk2OTI4ODcsImV4cCI6MTY4OTcyNTI4NywiYXVkIjoiYXBpLmRldi5zZWxmY2FyZS5wYWdvcGEuaXQiLCJpc3MiOiJTUElEIiwianRpIjoiXzk0ZDJmZTYyMDQ2NTUyODRjMGRjIn0.EDbsdpQgXlJSzyVgRqZy7yuUILe5FUlaerC3n1gv6SQrNvljXJxgm3GTv0912UQ6VV85e4oxGgc4LrcvpyLYZcgVe-5-2gNfbYNIPbIWqicaX4GPucQrSq47H0NEIaAv6-3qI2l1IhdH--72zUls_911RoAg_JdINr7em0vxy7wEoqjWOxgEsfQhEauT8oyRV6dDDied5zA9YQPy7a7KlhvI6juwS4sCdFnaonNzhBcZnqW4qzpec2NaAb1xJuHnnTp_tdMz6zExEhupeopmYdtIzYHUvxfohHr1L7eDRzi5RKEUfnRIldajfQEX_NUL9UTU4EkCcwUs-rSIKskQog

### CODE REVIEW

File: src/app/[locale]/_hooks/useLogin.tsx [hunk 1]
```diff
@@ -27,8 +27,8 @@ const useLogin = (): LoginData => {
   const logOut = () => {
     setIsLoggedIn(false);
     setUserLogged(undefined);
-    cookieTokenOps.delete();
-    cookieUserOps.delete();
+    storageTokenOps.delete();
+    storageUserOps.delete();
```
Comment (shadowsheep1) [inline src/app/[locale]/_hooks/useLogin.tsx:31]: Grouping this two functions in a more explicit single one would be more expressive? What do you think? e.g. `clearSessionData()`

File: src/app/[locale]/_hooks/useToken.tsx [hunk 2]
```diff
@@ -1,31 +1,37 @@
...
 
   useEffect(() => {
-    if (parseJwt(extractToken())) {
-      setToken(extractToken());
-      setTokenError('OK');
-      cookieTokenOps.write(extractToken());
-      cookieUserOps.write(userFromJwtToken(extractToken()));
-    } else {
-      setTokenError('ERROR');
+    if (typeof window !== 'undefined') {
```
Comment (shadowsheep1) [inline src/app/[locale]/_hooks/useToken.tsx:15]: You are using `typeof windows !== 'undefined'` many times. What do you think about writing a more epxlicit utility function?

**Review by shadowsheep1 — CHANGES_REQUESTED** [review]
A couple of suggestions ;)

File: src/app/[locale]/(pages)/logoutInit/page.tsx [hunk 3]
```diff
@@ -66,6 +66,7 @@ const Init = (): React.ReactElement => {
       <FAQ />
       <SelectIdp
         open={openDialog}
+        spidLevel="L1"
```
[the remark this answers is not in the document]
Comment (claudiodall) [reply src/app/[locale]/(pages)/logoutInit/page.tsx:69]: Here we expect to have L1 auth level -> https://pagopa.atlassian.net/wiki/spaces/IAEI/pages/706314633/Requisiti+portale+IO-ESCO#How%3F

File: src/app/[locale]/(pages)/validateSession/page.tsx [hunk 4]
```diff
@@ -1,21 +1,19 @@
...
-  const { tokenError } = useToken();
-
   useEffect(() => {
-    if (tokenError === 'ERROR') {
-      redirect(ROUTES.LOGOUT_AUTH_KO);
-    }
-    if (tokenError === 'OK') {
+    if (parseJwt(extractToken())) {
+      storageTokenOps.write(extractToken());
+      storageUserOps.write(userFromJwtToken(extractToken()));
```
[the remark this answers is not in the document]
Comment (claudiodall) [reply src/app/[locale]/(pages)/validateSession/page.tsx:13]: Yes, we are taking care of this in another branch. If is ok for you, you will see the changes in the next PR

[the remark this answers is not in the document]
Comment (shadowsheep1) [reply src/app/[locale]/(pages)/validateSession/page.tsx:13]: In this case we usually put a `// FIXME: Jira ticket number` to track the refactoring.

[the remark this answers is not in the document]
Comment (claudiodall) [reply src/app/[locale]/(pages)/validateSession/page.tsx]: Added the comment, thanks

File: src/app/[locale]/_component/selectIdp/idpList.tsx [hunk 5]
```diff
@@ -2,9 +2,15 @@ import React from 'react';
...
-export function IdpList() {
+interface Spid {
+  spidLevel: 'L1' | 'L2' | 'L3';
+}
+
+export function IdpList({ spidLevel }: Spid) {
   const getSPID = (IDP: IdentityProvider) => {
-    window.location.assign(`http://localhost:9090/login?entityID=${IDP.entityId}&authLevel=SpidL1`);
+    window.location.assign(
+      `http://localhost:9090/login?entityID=${IDP.entityId}&authLevel=Spid${spidLevel}`
```
[the remark this answers is not in the document]
Comment (claudiodall) [reply src/app/[locale]/_component/selectIdp/idpList.tsx:12]: Added 3 env and used env variable

File: src/app/[locale]/_hooks/useLogin.tsx (hunk 1 above)
Comment (claudiodall) [reply src/app/[locale]/_hooks/useLogin.tsx:31]: Updated with sessionStorage.clear();

File: src/app/[locale]/_hooks/useToken.tsx (hunk 2 above)
Comment (claudiodall) [reply src/app/[locale]/_hooks/useToken.tsx:15]: created isBrowser common utility function

File: src/app/[locale]/_component/selectIdp/selectIdp.tsx [hunk 6]
```diff
@@ -29,7 +34,7 @@ export function SelectIdp({ open, onClose }: IDialog) {
         >
           {'Scegli il tuo Identity Provider'}
         </Typography>
-        <IdpList />
+        <IdpList spidLevel={level} />
```
Comment (shadowsheep1) [inline src/app/[locale]/_component/selectIdp/selectIdp.tsx:37]:
```suggestion
        <IdpList spidLevel />
```

File: src/app/[locale]/(pages)/logoutInit/page.tsx [hunk 7]
```diff
@@ -64,12 +70,7 @@ const Init = (): React.ReactElement => {
         </Grid>
       </Grid>
       <FAQ />
-      <SelectIdp
-        open={openDialog}
-        onClose={(opn) => {
-          setOpenDialog(opn);
-        }}
-      />
+      <SelectIdp isOpen={openDialog} spidLevel={spidLevel} onClose={setOpenDialog} />
```
[the remark this answers is not in the document]
Comment (claudiodall) [reply src/app/[locale]/(pages)/logoutInit/page.tsx:73]: same as previous comment

File: src/app/[locale]/_component/selectIdp/selectIdp.tsx [hunk 8]
```diff
@@ -1,18 +1,23 @@
...
+export function SelectIdp({ isOpen, spidLevel, onClose }: IDialog) {
   const [openDialog, setOpenDialog] = useState<boolean>(false);
 
   useEffect(() => {
-    setOpenDialog(open);
-  }, [open]);
+    setOpenDialog(isOpen);
+  }, [isOpen]);
+
+  const level: SpidLevels = {
```
[the remark this answers is not in the document]
Comment (claudiodall) [reply src/app/[locale]/_component/selectIdp/selectIdp.tsx:18]: spidLevel is already declared as props, we cannot use duplicates, that's why we changed the name after the discriminated union on spidLavels

File: src/app/[locale]/(pages)/access/page.tsx [hunk 9]
```diff
@@ -154,6 +161,13 @@ const Access = (): React.ReactElement => {
           </Link>
         </Grid>
       </Grid>
+      <SelectIdp
+        isOpen={openDialog}
+        spidLevel={spidLevel}
```
[the remark this answers is not in the document]
Comment (shadowsheep1) [reply src/app/[locale]/(pages)/access/page.tsx:166]: Okay, you are right! It seems that property shorthand is working only for boolean types: https://github.com/facebook/jsx/issues/23. I didn't remember that.

**Comment by claudiodall** [comment]
After accepting the hint to use isBrowser() in useToken hook, I realized that the utility function wasn't automatically imported. 

After rebuilding the solution, I noticed that the useEffect in the session provider was being triggered too many times due to the various changes made. To avoid this problem, I had to change the dependency array items, we had also introduced pathName instead of window.location.
 
After testing, everything now seems to be working fine. 
Additionally, I took the opportunity to fix the authorization type in sessionProvider using a discriminated union, as advised.

I kindly request a final review before proceeding with the merge. Thank you!

**Review by shadowsheep1 — APPROVED** [review]
From a static code check POV, LGTM!

**Pull request merged**